march_joint_inertia_controller/src/inertia_estimator.cpp

```diff
@@ -0,0 +1,290 @@
+  std::vector<double> z = { 0.0, 0.0, 0.0 };
+  auto it = filtered_acceleration_array_.begin();
+  auto x_n = acceleration_array_[0];
+  double x;
+
+  for (unsigned int i = 0; i < (sizeof(sos_) / sizeof(sos_[0])); ++i)
+  {
+    x_n = acceleration_array_[0];
+    x = sos_[i][0] * x_n + z[0];
+    z[0] = sos_[i][1] * x_n - sos_[i][3] * x + z[1];
+    z[1] = sos_[i][2] * x_n - sos_[i][4] * x;
+
+    //    temp *= (acceleration_array_[0] * sos_[i][0] + acceleration_array_[1] * sos_[i][1] +
+    //             acceleration_array_[2] * sos_[i][2]) /
+    //            (acceleration_array_[0] * sos_[i][3] + acceleration_array_[1] * sos_[i][4] +
+    //             acceleration_array_[2] * sos_[i][5]);
+  }
+  it = filtered_acceleration_array_.begin();
+  it = filtered_acceleration_array_.insert(it, x);
+  filtered_acceleration_array_.resize(acc_size_);
+
+  z = { 0.0, 0.0, 0.0 };
+  for (unsigned int i = 0; i < (sizeof(sos_) / sizeof(sos_[0])); ++i)
+  {
+    x_n = joint_torque_[0];
+    x = sos_[i][0] * x_n + z[0];
+    z[0] = sos_[i][1] * x_n - sos_[i][3] * x + z[1];
+    z[1] = sos_[i][2] * x_n - sos_[i][4] * x;
+    //    temp *= (joint_torque_[0] * sos_[i][0] + joint_torque_[1] * sos_[i][1] + joint_torque_[2] * sos_[i][2]) /
+    //            (joint_torque_[0] * sos_[i][3] + joint_torque_[1] * sos_[i][4] + joint_torque_[2] * sos_[i][5]);
+  }
+  it = filtered_joint_torque_.begin();
+  it = filtered_joint_torque_.insert(it, x);
+  filtered_joint_torque_.resize(fil_tor_size_);
+
+  // temp = 1.0;
+}
+
+// Estimate the inertia using the acceleration and torque
+void InertiaEstimator::inertia_estimate()
+{
+  double torque_e;
+  double acc_e;
+  apply_butter();
+
+  correlation_calculation();
+  K_i_ = gain_calculation();
+  K_a_ = alpha_calculation();
+  torque_e = filtered_joint_torque_[0] - filtered_joint_torque_[1];
+  acc_e = filtered_acceleration_array_[0] - filtered_acceleration_array_[1];
+  joint_inertia_ = (torque_e - (acc_e * joint_inertia_)) * K_i_ * K_a_ + joint_inertia_;
+}
+
+// Calculate the alpha coefficient for the inertia estimate
+double InertiaEstimator::alpha_calculation()
+{
+  double vib = vibration_calculation();
+  if (vib < min_alpha_)
+  {
+    vib = min_alpha_;
+  }
+  else if (vib > max_alpha_)
+  {
+    vib = max_alpha_;
+  }
+  return (vib - min_alpha_) / (max_alpha_ - min_alpha_);
+}
+
+// Calculate the inertia gain for the inertia estimate
+double InertiaEstimator::gain_calculation()
+{
+  return (corr_coeff_ * (filtered_acceleration_array_[0] - filtered_acceleration_array_[1])) /
+         (lambda_ + corr_coeff_ * pow(filtered_acceleration_array_[0] - filtered_acceleration_array_[1], 2));
+}
+
+// Calculate the correlation coefficient of the acceleration buffer
+void InertiaEstimator::correlation_calculation()
+{
+  corr_coeff_ =
+      corr_coeff_ / (lambda_ + corr_coeff_ * pow(filtered_acceleration_array_[0] - filtered_acceleration_array_[1], 2));
+  double large_number = pow(10, 8);
```

Comment:
computing the power takes time, so you can convert this to a constant.
```suggestion
  const double large_number = 10e8;
```